liquidctl/driver/msi.py

```diff
@@ -0,0 +1,1468 @@
+        "breathing": _LightingMode.BREATHING,
+        "clock": _LightingMode.CLOCK,
+        "color pulse": _LightingMode.COLOR_PULSE,
+        "color ring": _LightingMode.COLOR_RING,
+        "color ring double flashing": _LightingMode.COLOR_RING_DOUBLE_FLASHING,
+        "color ring flashing": _LightingMode.COLOR_RING_FLASHING,
+        "color shift": _LightingMode.COLOR_SHIFT,
+        "color wave": _LightingMode.COLOR_WAVE,
+        "corsair ique": _LightingMode.CORSAIR_IQUE,
+        "disable": _LightingMode.DISABLE,
+        "disable2": _LightingMode.DISABLE2,
+        "double flashing": _LightingMode.DOUBLE_FLASHING,
+        "double meteor": _LightingMode.DOUBLE_METEOR,
+        "energy": _LightingMode.ENERGY,
+        "fan control": _LightingMode.FAN_CONTROL,
+        "fire": _LightingMode.FIRE,
+        "flashing": _LightingMode.FLASHING,
+        "jazz": _LightingMode.JAZZ,
+        "jrainbow": _LightingMode.JRAINBOW,
+        "lava": _LightingMode.LAVA,
+        "lightning": _LightingMode.LIGHTNING,
+        "marquee": _LightingMode.MARQUEE,
+        "meteor": _LightingMode.METEOR,
+        "movie": _LightingMode.MOVIE,
+        "msi marquee": _LightingMode.MSI_MARQUEE,
+        "msi rainbow": _LightingMode.MSI_RAINBOW,
+        "steady": _LightingMode.NO_ANIMATION,
+        "planetary": _LightingMode.PLANETARY,
+        "play": _LightingMode.PLAY,
+        "pop": _LightingMode.POP,
+        "rainbow": _LightingMode.RAINBOW,
+        "rainbow double flashing": _LightingMode.RAINBOW_DOUBLE_FLASHING,
+        "rainbow flashing": _LightingMode.RAINBOW_FLASHING,
+        "rainbow wave": _LightingMode.RAINBOW_WAVE,
+        "random": _LightingMode.RANDOM,
+        "rap": _LightingMode.RAP,
+        "stack": _LightingMode.STACK,
+        "visor": _LightingMode.VISOR,
+        "water drop": _LightingMode.WATER_DROP,
+        "end": _LightingMode.END,
+    }
+    BUILTIN_MODES = {
+        "silent": _FanMode.SILENT.value,
+        "balanced": _FanMode.BALANCE.value,
+        "game": _FanMode.GAME.value,
+        "default": _FanMode.DEFAULT.value,
+        "smart": _FanMode.SMART.value,
+    }
+    SCREEN_MODES = {
+        "hardware": _ScreenMode.HARDWARE,
+        "image": _ScreenMode.IMAGE,
+        "banner": _ScreenMode.BANNER,
+        "clock": _ScreenMode.CLOCK,
+        "settings": _ScreenMode.SETTINGS,
+        "disable": _ScreenMode.DISABLED,
+    }
+    HWMONITORDISPLAY = {
+        "cpu_freq": _OLEDHardwareMonitorOffset.CPU_FREQ,
+        "cpu_temp": _OLEDHardwareMonitorOffset.CPU_TEMP,
+        "gpu_freq": _OLEDHardwareMonitorOffset.GPU_MEMORY_FREQ,
+        "gpu_usage": _OLEDHardwareMonitorOffset.GPU_USAGE,
+        "fan_pump": _OLEDHardwareMonitorOffset.FAN_PUMP,
+        "fan_radiator": _OLEDHardwareMonitorOffset.FAN_RADIATOR,
+        "fan_cpumos": _OLEDHardwareMonitorOffset.FAN_CPUMOS,
+    }
+    _MATCHES = [
+        (0x0DB0, 0xB130, "MSI MPG Coreliquid K360", {"fan_count": 5}),
+        (0x0DB0, 0xCA00, "Unknown", {}),
```

Comment:
Should they be removed (commented out for future contributions) if it is undesirable to have technically unsupported devices in the udev rules? These were added in the original PR, to start I now at least wrote the more descriptive names.

Reply:
How confident are we that those suspected IDs will really work properly with the driver?

Depending on that it might be wise to either comment them out (if not much) or leave them in but require an [unsafe flag](https://github.com/liquidctl/liquidctl/blob/e3f91793b35550f1b454e26e37f110c70911b9eb/liquidctl/driver/ddr4.py#L212-L215) for actual use (if quite confident).

---

Please suffix all new device descriptions with ` (experimental)`.

Reply:
I am reasonably confident that the driver should be mostly functional for the listed devices, as there are no meaningful differences in the communication protocol compared to the decompiled manufacturer implementation. [Some issues](https://github.com/liquidctl/liquidctl/pull/564#issuecomment-1675871831) clearly still occur, but from that it seems that the fan control does work on the S360. I added unsafe checks for this.